### DISCUSSION TITLE
@JsonSetter (nulls = Nulls.AS_EMPTY)

### DISCUSSION THREAD

**Discussion by ghost** [FasterXML/jackson-databind#3821] [Q&A]
@JsonSetter (nulls = Nulls.AS_EMPTY)
becomes ""
How to assign an enumeration
now protocol is null
I want to make "" become an enumeration

@JsonSetter(nulls = Nulls.AS_EMPTY)
private Constants.PROTO_TYPE protocol;

**Comment by cowtowncoder** [comment]
What exactly are you doing? What is the input, what kind of type is `Constants`? What becomes ""?

You are missing a lot of relevant information here to try to help.

**Reply by ghost** [reply]
@JsonSetter(nulls = Nulls.AS_EMPTY)
private Constants.PROTO_TYPE protocol;

when input {"protocol":null}
the enum Constants.PROTO_TYPE  not accept
ask @JsonSetter  any other way handle this question

**Reply by cowtowncoder** [reply]
Unfortunately that is no more useful information. Sorry, I can not really help here.

**Comment by Raphasha27** [comment]
The `Nulls.AS_EMPTY` setting tells Jackson to use the empty/default value for a type when the input is null, but you need to pair it with `@JsonSetter` on the setter, not the field:

```java
public class Item {
    private String name;
    private List<String> tags;

    @JsonSetter(nulls = Nulls.AS_EMPTY)
    public void setTags(List<String> tags) {
        this.tags = tags != null ? tags : Collections.emptyList();
    }
}
```

For this to work, the setter must already handle the empty case. The `nulls` attribute only tells Jackson to skip calling the setter (using the empty value from constructor/default) or pass the null. 

If you want globally empty collections for null fields, configure the ObjectMapper:

```java
ObjectMapper mapper = new ObjectMapper();
mapper.setDefaultSetterInfo(JsonSetter.Value.forValueNulls(Nulls.AS_EMPTY));
```

Note: `Nulls.AS_EMPTY` requires Jackson 2.9+. It works for collections, strings, and optionals — for custom types you need to define what "empty" means via `@JsonFormat(with = JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY)` or a custom deserializer.
